### PR TITLE
Minor: make memory_limit tests more self describing

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/5885

## Rationale for this change
As DataFusion's memory limiting gets more sophisticated so do the tests for these limits. I made them better in https://github.com/apache/arrow-datafusion/pull/7131 and https://github.com/apache/arrow-datafusion/pull/7130 but they could be more self describing

## What changes are included in this PR?
Complete switching memory_limit tests to builder style so each test reads more fluently and it is easier to understand what is going on

### CODE REVIEW

File: datafusion/core/tests/memory_limit.rs [hunk 1]
```diff
@@ -55,118 +55,110 @@ fn init() {
 
 #[tokio::test]
 async fn oom_sort() {
-    TestCase::new(
-        "select * from t order by host DESC",
-        vec![
+    TestCase::new()
+        .with_query("select * from t order by host DESC")
```
Comment (alamb) [inline datafusion/core/tests/memory_limit.rs:59]: the major change is that by using a builder pattern, now the tests are more explicit about what the values are

**Review by metesynnada — COMMENTED** [review]
LGTM, thanks for the effort.

**Comment by alamb** [comment]
Thank you for the review @metesynnada

**Review by yjshen — APPROVED** [review]
After the refactor, the builder-style tests appear much clearer; thanks @alamb!

**Pull request merged**